sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/HadoopFsRelation.scala

```diff
@@ -49,9 +49,12 @@ case class HadoopFsRelation(
   override def sqlContext: SQLContext = sparkSession.sqlContext
 
   val schema: StructType = {
-    val dataSchemaColumnNames = dataSchema.map(_.name.toLowerCase).toSet
-    StructType(dataSchema ++ partitionSchema.filterNot { column =>
-      dataSchemaColumnNames.contains(column.name.toLowerCase)
+    val equality = sparkSession.sessionState.conf.resolver
+    val overriddenDataSchema = dataSchema.map { dataField =>
```

Comment:
Because the current code will iterate `dataSchema` many times(depend on the number of partition columns), while my proposal only iterate it 2 times.